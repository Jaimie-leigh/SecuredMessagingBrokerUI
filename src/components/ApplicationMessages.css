.detailsBanner {
  background-color: whitesmoke;
  display: flex;
  padding: 0.5em;
  align-items: center;
}

.rollNumber {
  flex: 1;
}

.returnToApps {
  flex: 1;
  text-align: end;
}

.returnToApps button {
  min-width: auto;
  margin: 0em;
}

.appMessageMainContent {
  margin-left: 10%;
  margin-right: 10%;
  margin-bottom: 5%;
  padding: 0.5em;
}

.appMessageMainContent h1 {
  color: #071d49;
}

.formHeading {
  margin: 0em;
  margin-bottom: 0.5em;
}

.messageSubjectHeader {
  background-color: #edf6ff;
  color: #071d49;
  display: flex;
  align-items: center;
  font-weight: bold;
  padding: 0.5em;
  border-top: solid 2px #005eb8;
  border-bottom: solid 1px #005eb8;
}

.messageSubject {
  background-color: white;
  display: flex;
  align-items: center;
  font-weight: bold;
  padding: 0.5em;
  border-bottom: solid 1px #b5bbc8;
}

.messageSubject:last-child {
  border-bottom: solid 1px #005eb8;
}

.questionIcon {
  width: 4em;
  height: 4em;
}

.innerMS {
  flex: 1;
}

.innerMS:last-child {
  text-align: right;
}

.innerMSDownIcon {
  width: 1.5em;
  height: 1.5em;
}

.chatIcon {
  width: 3em;
  height: 3em;
  padding: 1em;
}

.sendMessageContent {
  display: flex;
  margin: 2em;
  margin-left: 10%;
  margin-right: 10%;
}

.sentMessageText {
  background-color: #00a2b3;
  color: white;
  border-radius: 15px;
  padding: 1em;
  padding-top: 2em;
  text-align: center;
  margin-right: 2em;
  width: 100%;
}

p {
  font-size: medium;
}

.sendMessageForm {
  padding: 1.5em;
  border-radius: 6px;
  margin: auto;
  flex-direction: column;
  box-shadow: 0px 8px 36px #222;
  max-width: unset;
  width: 100%;
}

.form-group {
  margin-bottom: 0.5em;
}

.form-group input {
  margin-bottom: 0.5em;
}

.sendMessageForm input {
  flex: auto;
  max-width: max-content;
  min-width: -webkit-fill-available;
}

.sendMessageForm select {
  flex: auto;
  width: auto;
  min-width: -webkit-fill-available;
  height: max-content;
}

.formRadio {
  display: inline-flex;
}

.formRadiob {
  margin-right: 3em;
}

.formRadio label {
  margin-left: 0.5em;
}

.formRadio input {
  width: auto;
  display: inline;
}

.sendMessageForm textarea {
  padding: 10px 10px;
  border-radius: 5px;
  border: 1px solid #d6d1d5;
  margin-top: 5px;
  display: block;
  width: -webkit-fill-available;
}

.messageHistoryText {
  margin-bottom: 2em;
}

.showHideIcons {
  width: 2em;
  height: 2em;
}

.innerMS button {
  min-width: fit-content;
  background-color: white;
}

button:focus {
  outline: 0;
}

.messageHistoryHeader {
  background-color: #edf6ff;
  color: #071d49;
  display: flex;
  align-items: center;
  font-weight: bold;
  padding: 0.5em;
  border-top: solid 2px #005eb8;
  border-bottom: solid 1px #005eb8;
}

.messageHistoryBody {
  background-color: white;
  display: flex;
  align-items: center;
  padding: 0.5em;
  border-bottom: solid 1px #b5bbc8;
}

.messageHistoryBody:last-child {
  border-bottom: solid 1px #005eb8;
}

.innerMS-history-body {
  flex: 2;
}
