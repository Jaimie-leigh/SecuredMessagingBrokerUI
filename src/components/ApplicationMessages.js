import React from "react";
import "./ApplicationMessages.css";
import "./loading.css";
import FadeLoader from "react-spinners/FadeLoader";
import { ReactComponent as DownIcon } from "../images/down-chevron.svg";
import { ReactComponent as UpIcon } from "../images/up-chevron.svg";
import { ReactComponent as ChatIcon } from "../images/chat.svg";
import { ReactComponent as QuestionIcon } from "../images/question.svg";
import Moment from "moment";
import { Form, Button } from "react-bootstrap";

class ApplicationMessages extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      data: [],
      messageChainData: [],
      messageSubjectData: [],
      isLoading: false,
      checkboxes: [],
      selectedId: [],
      formLableSelected: "new",
      formSelectedSubject: "",
      formSelectedSubjectId: "",
      formNewSubject: "",
      formChainID: "",
      formMessageBody: "",
      messageSent: false,
      brokerID: props.location.state.brokerID,
    };
    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  componentDidMount() {
    this.setState({ isLoading: true });
    // gets all data for spesific application based on roll number
    // const proxyurl = "https://cors-anywhere.herokuapp.com/";
    const jaiProxyURL = "https://nameless-hamlet-06819.herokuapp.com/";
    const url =
      "http://securedmessaging.azurewebsites.net/api/ApplicationMessages/" +
      this.props.location.state.rollNumber;
    fetch(jaiProxyURL + url)
      .then((res) => res.json())
      .then((data) => this.setState({ data: data }))
      .then(
        fetch(
          jaiProxyURL +
            "http://securedmessaging.azurewebsites.net/api/Message_Chain"
        )
          .then((res) => res.json())
          .then((messageChainData) =>
            this.setState({ messageChainData: messageChainData }, () => {})
          )
      )
      .then(
        // gets all message_subject to determin the next subject ID when user is sending new message
        fetch(
          jaiProxyURL +
            "http://securedmessaging.azurewebsites.net/api/Message_Subject"
        )
          .then((res) => res.json())
          .then((messageSubjectData) =>
            this.setState({
              messageSubjectData: messageSubjectData,
              isLoading: false,
            })
          )
      );
  }

  handleClick = (id) => {
    if (this.state.selectedId !== id) {
      this.setState({ selectedId: id });
    } else {
      this.setState({ selectedId: null });
    }
  };

  //handle change subject sets the state of formSelectedSubject and formSelectedSubjectId
  // these will be needed in the POST to api
  handleChangeSubject = (event) => {
    this.setState({ formSelectedSubject: event.target.value }, () => {
      //callback
      this.state.data.message_Subjects.map((ms) => {
        if (ms.subject === this.state.formSelectedSubject) {
          this.setState({ formSelectedSubjectId: ms.messageSubjectId }, () => {
            //callback to set state instantly
          });
          return null;
        } else {
          return null;
        }
      });
    });
  };

  handleBodyChange = (event) => {
    this.setState({ formMessageBody: event.target.value });
  };

  handleNewSubject = (event) => {
    this.setState({ formNewSubject: event.target.value });
  };

  // handles user submiting the message from and sends post request to API
  // show loading screen when message submit. sends message to API and reloads page to display message in message history
  handleFormSubmit(e) {
    this.setState({ isLoading: true });
    e.preventDefault();

    // set next chainID
    const count = this.state.messageChainData.length - 1;
    // get last message in chain and its ID, then +1 for next new chain id
    const chainID = this.state.messageChainData[count].messageChainId + 1;
    this.setState({ formChainID: chainID }, () => {
      //callback to set state instantly
    });
    // get current time to add to post body 'date and time'
    let submit_time = Moment().format("ddd DD MMM YYYY HH:mm:ss");;
    // set message subject from state, will change in if statement if new message
    let messageSubjectId = this.state.formSelectedSubjectId;

    //const proxyurl = "https://cors-anywhere.herokuapp.com/";
    const jaiProxyURL = "https://nameless-hamlet-06819.herokuapp.com/";

    // IF replying to exisiting message NEW subject DOES NOT need creating first
    // If NEW suject two fetch methods one for subject then one for chain
    if (this.state.formLableSelected === "new") {
      messageSubjectId = this.state.messageSubjectData.length + 1;

      try {
        fetch(
          jaiProxyURL +
            "http://securedmessaging.azurewebsites.net/api/Message_Subject",
          {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            // convert the React state to JSON and send it as the POST body
            body: JSON.stringify({
              messageSubjectId: messageSubjectId,
              subject: this.state.formNewSubject,
              rollNumber: this.props.location.state.rollNumber,
              BrokerID: parseInt(this.state.brokerID),
              dateTime: submit_time,
            }),
          }
        ).then(
          fetch(
            jaiProxyURL +
              "http://securedmessaging.azurewebsites.net/api/Message_Chain",
            {
              method: "POST",
              headers: { "Content-Type": "application/json" },
              // convert the React state to JSON and send it as the POST body
              body: JSON.stringify({
                messageChainId: chainID,
                messageSubjectId: messageSubjectId,
                messageBody: this.state.formMessageBody,
                sentFromId: parseInt(this.state.brokerID),
                dateTime: submit_time,
              }),
            }
          ).then((data) =>
            this.setState(
              { messageSent: true, isLoading: false, formLableSelected: "new" },
              this.componentDidMount()
            )
          )
        );
      } catch (error) {}
    } else {
      try {
        fetch(
          jaiProxyURL +
            "http://securedmessaging.azurewebsites.net/api/Message_Chain",
          {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            // We convert the React state to JSON and send it as the POST body
            body: JSON.stringify({
              messageChainId: chainID,
              messageSubjectId: messageSubjectId,
              messageBody: this.state.formMessageBody,
              sentFromId: parseInt(this.state.brokerID),
              dateTime: submit_time,
            }),
          }
        ).then((data) =>
          this.setState(
            { messageSent: true, isLoading: false, formLableSelected: "new" },
            this.componentDidMount()
          )
        );
      } catch (error) {}
    }
  }

  render() {
    const { data, isLoading } = this.state;

    if (isLoading) {
      return (
        <div className="pageLoading">
          <p>Loading...</p>
          <FadeLoader size={150} color={"#005EB8"} loading={isLoading} />
        </div>
      );
    }

    if (data.length === 0) {
      return <p> no data found</p>;
    }

    console.log(data);
    return (
      <div>
        <div className="detailsBanner">
          <div className="rollNumber">
            {" "}
            <strong>
              Application roll number: {this.props.location.state.rollNumber}
            </strong>
          </div>
          <div className="returnToApps">
            <button onClick={this.props.history.goBack}>
              Return to All Applications
            </button>
          </div>
        </div>
        {this.state.messageSent === true && (
          <div className="messageSent">{this.displaySentText()}</div>
        )}
        <div className="sendMessageContent">
          <div className="sentMessageText">
            <QuestionIcon className="questionIcon" />
            <h1>Need to ask us a question about this application?</h1>
            <h2>We're here to help</h2>
            <p>
              You can speak to us from the comfort of your own home, at your own
              convienance. Start a new chat or continue an exisiting chat using
              the form on the right and we will get back to you as soon as
              possible.
            </p>
            <p>
              All message chats for this application can be found in the
              <strong>"Application Chat History"</strong> section below.
            </p>
            <p>
              We are open Monday to Friday 8am to 6pm and Saturday 9am to 4pm.
              Weâ€™re closed Sundays and Bank Holidays.
            </p>
          </div>
          <Form className="sendMessageForm" onSubmit={this.handleFormSubmit}>
            <div>
              <ChatIcon className="chatIcon" />
              <h1 className="formHeading">
                Start a new chat or reply to an existing chat
              </h1>
            </div>
            <Form.Group>
              <Form.Label>
                Roll Number
                <Form.Control
                  type="text"
                  placeholder={this.props.location.state.rollNumber}
                  disabled
                />
              </Form.Label>
            </Form.Group>
            <Form.Group className="formRadio">
              <Form.Check
                className="formRadiob"
                type="radio"
                label="New chat"
                value="new"
                name="neworexisting"
                id="New Message"
                onChange={this.onFormMessageChanged}
                defaultChecked
              />
              <Form.Check
                className="formRadiob"
                type="radio"
                label="Reply to existing  chat"
                value="reply"
                name="neworexisting"
                id="existing Message"
                onChange={this.onFormMessageChanged}
              />
            </Form.Group>
            {this.returnCorrectFormFields(data)}
            <Form.Group>
              <Form.Label>
                Message Body
                <Form.Control
                  as="textarea"
                  rows={3}
                  value={this.state.value}
                  onChange={this.handleBodyChange}
                  required
                  maxLength="250"
                  minLength="10"
                />
              </Form.Label>
            </Form.Group>
            <Button variant="primary" type="submit">
              Send Message
            </Button>
          </Form>
        </div>
        <div className="appMessageMainContent">
          <div className="messageHistoryText">
            <h1>Application Chat History</h1>
            <p>
              Here you can find all sent and received chats for this
              Application.
            </p>
            {/* <p>
              To expand a set of messages to see more detail including, the
              message content, message date and time and who sent the message,
              press the down icon that looks like this:{" "}
              <DownIcon className="innerMSDownIcon" />.
            </p> */}
            <p>
              To continue any existing chat's use the form above and select
              'Reply to existing chat' and we will get back to you as soon as we
              can. Alternative you can start a New chat if you need to ask us
              something new.
            </p>
          </div>
          <div className="messageSubjectHeader">
            <div className="innerMS">Chat Subject</div>
            <div className="innerMS">Number of messages in chat</div>
            <div className="innerMS">Latest message Date and Time</div>
            <div className="innerMS">View chat history</div>
          </div>
          {data?.message_Subjects?.map((ms, index) => (
            <div key={index + ms.toString()}>
              <div className="messageSubject">
                <div className="innerMS">{ms.subject}</div>
                <div className="innerMS">{ms.message_Chain.length}</div>
                <div className="innerMS">
                  {this.getLatestMessageDateTime(ms.message_Chain)}
                </div>
                <div className="innerMS">
                  <button onClick={() => this.handleClick(ms.messageSubjectId)}>
                    {this.state.selectedId === ms.messageSubjectId ? (
                      <UpIcon className="showHideIcons" />
                    ) : (
                      <DownIcon className="showHideIcons" />
                    )}
                  </button>
                </div>
              </div>
              {this.state.selectedId === ms.messageSubjectId && (
                <div className="messageHistory">
                  {this.getAllMessageInChain(ms.message_Chain)}
                </div>
              )}
            </div>
          ))}
        </div>
      </div>
    );
  }

  displaySentText() {
    return (
      <div>
        <p>Your message has been sent!</p>
      </div>
    );
  }

  returnToApp() {
    this.props.history.goBack();
  }

  returnCorrectFormFields(data) {
    if (this.state.formLableSelected !== "new") {
      return this.choseMessageSubject(data);
    } else {
      return this.newMessageSubject(data);
    }
  }

  choseMessageSubject(data) {
    return (
      <Form.Group>
        <Form.Label>
          Select the message subject
          <Form.Control as="select" onChange={this.handleChangeSubject}>
            <option value="0">Choose...</option>
            {data.message_Subjects.map((ms, index) => (
              <option value={ms.subject} key={ms.toString() + index}>
                {ms.subject}
              </option>
            ))}
          </Form.Control>
        </Form.Label>
      </Form.Group>
    );
  }

  newMessageSubject(data) {
    return (
      <Form.Group>
        <Form.Label>Enter Message Subject</Form.Label>
        <Form.Control
          type="text"
          placeholder="Enter message subject"
          value={this.state.value}
          onChange={this.handleNewSubject}
          required
          maxLength="100"
          minLength="4"
        />
      </Form.Group>
    );
  }

  onFormMessageChanged = (event) => {
    this.setState({
      formLableSelected: event.target.value,
    });
  };

  getAllMessageInChain(messageChain) {
    return (
      <div className="messageHistory">
        <div className="messageHistoryHeader">
          <div className="innerMS-history-body">Message</div>
          <div className="innerMS">Date and Time</div>
          <div className="innerMS">Message sent by</div>
        </div>
        {messageChain.map((ms, index) => (
          <div className="messageHistoryBody" key={ms.toString() + index}>
            <div className="innerMS-history-body">{ms.messageBody}</div>
            <div className="innerMS">
              {Moment(ms.dateTime).format("ddd DD MMM YYYY HH:mm")}
            </div>
            <div className="innerMS">{ms.sentFromId}</div>
          </div>
        ))}
      </div>
    );
  }

  getLatestMessageDateTime(messageChain) {
    const lastmessage = messageChain.length - 1;

    Moment.locale("en");
    try {
      var dt = messageChain[lastmessage].dateTime;
      return Moment(dt).format("ddd DD MMM YYYY HH:mm");
    } catch (error) {
      return "Information not avalible";
    }
  }
}

export default ApplicationMessages;
