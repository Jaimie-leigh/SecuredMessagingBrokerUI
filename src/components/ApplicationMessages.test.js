import React from "react";
import renderer from "react-test-renderer";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import ApplicationMessages from "./ApplicationMessages";
import { createMemoryHistory } from "history";

let container = null;

beforeEach(() => {
  // set a DOM elemant as render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // clean up on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("ApplicationMessages tests", () => {
  test("test renders as per snapshot", () => {
    const state = {
      rollNumber: 1000000001,
      brokerID: 1234567,
    };
    const tree = renderer
      .create(
        <ApplicationMessages location={{ state: state }}></ApplicationMessages>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("test page renders as epxected when loading", () => {
    const state = {
      rollNumber: 1000000001,
      brokerID: 1234567,
    };
    render(
      <ApplicationMessages location={{ state: state }}></ApplicationMessages>,
      container
    );
    expect(container.innerHTML).toContain("Loading...");
  });

  test("mock fetch call, empty responce from db should return no data", async () => {
    const fakeResponse = [];
    const state = {
      rollNumber: 1000000001,
      brokerID: 1234567,
    };

    jest.spyOn(window, "fetch").mockImplementation(() => {
      const fetchResponse = {
        json: () => Promise.resolve(fakeResponse),
      };
      return Promise.resolve(fetchResponse);
    });

    await act(async () => {
      render(<ApplicationMessages location={{ state: state }} />, container);
    });

    expect(container.innerHTML).toContain("no data");

    window.fetch.mockRestore();
  });

  test("mock fetch call, valid responce from db should return applications", async () => {
    const fakeResponse = [
      {
        message_Subjects: [
          {
            messageSubjectId: "1",
            brokerId: "1000001",
            subject: "Verify Documents",
            rollNumber: "1000000001",
            dateTime: "2020-09-29T15:16:00",
          },
          {
            messageSubjectId: "2",
            brokerId: "1000002",
            subject: "Valuation Delayed",
            rollNumber: "1000000002",
            dateTime: "2020-09-29T16:00:00",
          },
        ],
      },
    ];

    const state = {
      rollNumber: 1000000001,
      brokerID: 1234567,
    };
    const history = createMemoryHistory();

    jest.spyOn(window, "fetch").mockImplementation(() => {
      const fetchResponse = {
        json: () => Promise.resolve(fakeResponse),
      };
      return Promise.resolve(fetchResponse);
    });

    await act(async () => {
      render(
        <ApplicationMessages history={history} location={{ state: state }} />,
        container
      );
    });

    expect(container.innerHTML).toContain(
      "Need to ask us a question about this application?", "Application Chat History"
    );

    window.fetch.mockRestore();
  });
});
