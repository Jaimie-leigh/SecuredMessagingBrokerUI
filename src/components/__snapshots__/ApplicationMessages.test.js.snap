// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationMessages tests test renders as per snapshot 1`] = `
<div
  className="pageLoading"
>
  <p>
    Loading...
  </p>
  <div
    className="css-1ujetdd"
  >
    <div
      className="css-137o4g7"
    />
    <div
      className="css-1jixpcn"
    />
    <div
      className="css-105wkyy"
    />
    <div
      className="css-1hljb6y"
    />
    <div
      className="css-gb69iu"
    />
    <div
      className="css-1t94dra"
    />
    <div
      className="css-tk7sl6"
    />
    <div
      className="css-ixmspk"
    />
  </div>
</div>
`;
